### PR TITLE
feat: [HL-33`] Reusable button loading indicator  🥛🍫

### PR DESCRIPTION
**Pull request by oradwanomar**
#### Proposed Changes

Add logic to UIButton extension

#### Additional Information

``` swift
let button = UIButton()
button.setTitle("Title", for: .normal)

button.startLoading()
// perform loading
button.stopLoading()
```

#### Screenshot

Same as here 😁 https://github.com/motoon-eg/healthy/pull/75

#### Icons Catalog

- 🍕 Feature
- 🐛 Bug Fix
- 📝 Documentation Update
- 🎨 Style
- 🧑‍💻 Code Refactor
- 🔥 Performance Improvements
- ✅ Test
- 🤖 Build
- 🔁 CI
- 📦 Chore (Release)
- ⏩ Revert
-->

### CODE REVIEW

File: Healthy/Classes/Utilities/Animatable.swift [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        return subviews.first(where: { $0 is UIActivityIndicatorView }) as? UIActivityIndicatorView
+    }
+    
+    /// Start animating while waiting another task is done
+    /// We use an associated object to store the original button title when startLoading() is called.
+    /// - Parameter color: indicator color
+    func startAnimation(withColor color: UIColor = .black) {
+        let originalTitle = title(for: .normal)
+        isEnabled = false
+        setTitle("", for: .normal)
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Animatable.swift:29]: We lost the title, forever 😅

Comment (ahmdmhasn) [reply Healthy/Classes/Utilities/Animatable.swift:29]: I think we can access titleLabel and update it's alpha instead?!

File: Healthy/Classes/Utilities/Animatable.swift [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        self.setupIndicatorLayout()
+        activityIndicator?.color = color
+        activityIndicator?.startAnimating()
+        objc_setAssociatedObject(self,
+                                 AssociatedKeys.originalTitle,
+                                 originalTitle,
+                                 .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Animatable.swift:36]:
```suggestion
        objc_setAssociatedObject(self,
                                 AssociatedKeys.originalTitle,
                                 originalTitle,
                                 .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
```

How to use it without associated object? 

```swift
private final class ButtonActivityIndicatorView: UIActivityIndicatorView {}
```

and get the first subview with this type?!

File: Healthy/Classes/Utilities/Animatable.swift [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    /// Start animating while waiting another task is done
+    /// - Parameter color: indicator color
+    func startAnimation(withColor color: UIColor)
+    
+    /// Stop animating when the task is finished
+    func stopAnimation()
+}
+
+/// Default implementation to UIButton loading indicator
+extension Animatable where Self: UIButton {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Animatable.swift:16]:
```suggestion
extension UIButton: Animatable {
```

**Review by ahmdmhasn — APPROVED** [review]
Thanks for your time ❤️
Please feel free to discuss the comments...

File: Healthy/Classes/Utilities/Animatable.swift (hunk 1 above)
Comment (oradwanomar) [reply Healthy/Classes/Utilities/Animatable.swift:29]: > We lost the title, forever 😅

I save it to original title with associated object 
But yes I want to discuss with you a way to save our title that more efficient than objc dynamic mechanism
This solution is working fine but I think there another way to enhance

Comment (oradwanomar) [reply Healthy/Classes/Utilities/Animatable.swift:29]: > I think we can access titleLabel and update it's alpha instead?!

That's a great solution and make more sense.
Thanks ya ahmed ❤️🚀

File: Healthy/Classes/Utilities/Animatable.swift (hunk 2 above)
Comment (oradwanomar) [reply Healthy/Classes/Utilities/Animatable.swift:36]: We need to discuss & try this

**Comment by ahmdmhasn** [comment]
As agreed, we will proceed with https://github.com/motoon-eg/healthy/pull/75, thanks @oradwanomar for your efforts! ❤️

**Pull request closed without merging**